libraries/Filter/SlewLimiter2D.cpp

```diff
@@ -0,0 +1,120 @@
+#define DERIVATIVE_CUTOFF_FREQ 25.0f
+
+SlewLimiter2D::SlewLimiter2D(const float &_slew_rate_max, const float &_slew_rate_tau) :
+    slew_rate_max(_slew_rate_max),
+    slew_rate_tau(_slew_rate_tau)
+{
+    slew_filter.set_cutoff_frequency(DERIVATIVE_CUTOFF_FREQ);
+    slew_filter.reset();
+}
+
+/*
+  apply filter to sample, returning multiplier between 0 and 1 to keep
+  output within slew rate
+ */
+float SlewLimiter2D::modifier(const Vector2f& sample, float dt)
+{
+    if (!is_positive(dt)) {
+        return 1.0;
+    }
+
+    // Calculate a low pass filtered slew rate
+    const float slew_rate = slew_filter.apply((sample - last_sample) / dt, dt).length();
+    last_sample = sample;
+
+    uint32_t now_ms = AP_HAL::millis();
+
+    // Apply a filter to decay maximum seen slew rate once the value had left the window period
+    const float decay_alpha = fminf(dt, slew_rate_tau) / slew_rate_tau;
+
+    // Apply a filter to increases in slew rate only to reduce the effect of gusts and large controller setpoint changes
+    const float attack_alpha = fminf(2.0f * decay_alpha, 1.0f);
+
+    // Decay the peak positive and negative slew rate if they are outside the window
+    // Never drop PID gains below 10% of configured value
+    if (slew_rate > _max_slew_rate) {
+        _max_slew_rate = slew_rate;
+        _max_slew_event_ms = now_ms;
+    } else if (now_ms - _max_slew_event_ms > WINDOW_MS) {
+        _max_slew_rate *= (1.0f - decay_alpha);
+    }
+
+    const float raw_slew_rate = _max_slew_rate;
+    _output_slew_rate = (1.0f - attack_alpha) * _output_slew_rate + attack_alpha * raw_slew_rate;
+    _output_slew_rate = fminf(_output_slew_rate, raw_slew_rate);
+
+    // skip calculation of gain modifier to avoid floating point exception
+    if (slew_rate_max <= 0) {
+        return 1.0;
+    }
+
+    // Constrain slew rate used for calculation
+    const float limited_raw_slew_rate = fminf(_max_slew_rate, 10.0f * slew_rate_max);
+
+    // Store a series of positive slew rate exceedance events
+    if (slew_rate > slew_rate_max) {
+        if (_event_index >= SLEWLIMITER2D_N_EVENTS) {
+            _event_index = 0;
+        }
+        _event_ms[_event_index] = now_ms;
+        _event_index++;
+    }
+
+    // Find the oldest event time
+    uint32_t oldest_ms = now_ms;
+    for (uint8_t index = 0; index < SLEWLIMITER2D_N_EVENTS; index++) {
+        if (_event_ms[index] > 0) {
+            oldest_ms = MIN(oldest_ms, _event_ms[index]);
+        }
+    }
+
+    // Apply a further reduction when the oldest exceedance event falls outside the window required for the
+    // specified number of exceedance events. This prevents spikes due to control mode changed, etc causing
+    // unwanted gain reduction and is only applied to the slew rate used for gain reduction
+    float modifier_input = limited_raw_slew_rate;
+    if (now_ms - oldest_ms > (SLEWLIMITER2D_N_EVENTS + 1) * WINDOW_MS) {
```

Comment:
Paul Riseborough says this WINDOW_MS is probably too low.

Reply:
I've increased the WINDOW_MS to 1000ms meaning the detector works with oscillations as low as 0.5hz.